Refactor Duplicate Code Patterns in StaticScope Class

danielathome19 opened the pull request:
These changes are aimed at improving code readability and maintainability within the `StaticScope` class of the AST folder.

- **Optimized Print and PrintID Methods:**
    Introduced a template function `PrintCommon` to handle common logic in `Print` and `PrintID`. This change simplifies the class interface and avoids repetition, enhancing code readability.

- **Simplified TryResolveHere Method:**
    Streamlined the `TryResolveHere` method by simplifying the conditional logic. The refactored code is more readable and easier to understand, improving overall code quality.

danielathome19 committed:
Refactored StaticScope for maintainability and readability.

1. Combined MarkDeclared and MarkUsable into UpdateNameStatus to reduce duplication.
2. Created PrintCommon template for shared logic in Print and PrintID.
3. Simplified logic in TryResolveHere for better readability.

chandlerc commented:
Just FYI, we've *just* started pivoting our focus much more towards the toolchain rather than the explorer: https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p3532.md

We're still updating documentation to make this clear, sorry if you happen to have dropped in right in-between this changing a bit.

That said, somewhat up to @zygoloid whether he thinks it makes sense to review and merge this, but it may be a bit slow / lower priority for folks.

Would love to have contributions to improve our docs around this and on the toolchain code itself!

zygoloid commented on explorer/ast/static_scope.cpp:
```diff
@@ -35,43 +35,43 @@ auto StaticScope::Add(std::string_view name, ValueNodeView entity,
+  it->second.status = new_status;
+  if (trace_stream_->is_enabled()) {
+    trace_stream_->Result() << trace_message << "\n";
   }
 }
 
+void StaticScope::MarkDeclared(std::string_view name) {
+  UpdateNameStatus(name, NameStatus::DeclaredButNotUsable, 
+                   "marked `" + std::string(name) + 
+                   "` declared but not usable in `" + PrintAsID(*this) + "`");
+}
+
 void StaticScope::MarkUsable(std::string_view name) {
-  auto it = declared_names_.find(name);
-  CARBON_CHECK(it != declared_names_.end()) << name << " not found";
-  it->second.status = NameStatus::Usable;
-  if (trace_stream_->is_enabled()) {
-    trace_stream_->Result()
-        << "marked `" << name << "` usable in `" << PrintAsID(*this) << "`\n";
-  }
+  UpdateNameStatus(name, NameStatus::Usable, 
+                   "marked `" + std::string(name) + 
+                   "` usable in `" + PrintAsID(*this) + "`");
```
I would prefer to not make this portion of the change, so that we don't build the error message unnecessarily in the common case where the call is successful.

danielathome19 replied:
Not a problem -- I've committed the reversion and updated the PR description for clarity.

zygoloid reviewed (commented):
The `Print` and `TryResolveHere` changes look like improvements to me. I'd be happy for us to land those parts since you've already put together the PR, but in general we're moving away from explorer development, so I wouldn't encourage spending more time on refactoring work here.

danielathome19 committed:
Update static_scope.cpp -- Removed refactoring of MarkUsable and MarkDeclared methods

Reverted the `MarkUsable` and `MarkDeclared` methods to the original state -- removed `UpdateNameStatus` helper method.

danielathome19 commented:
> The `Print` and `TryResolveHere` changes look like improvements to me. I'd be happy for us to land those parts since you've already put together the PR, but in general we're moving away from explorer development, so I wouldn't encourage spending more time on refactoring work here.

Sounds great! I'll keep that in mind.

danielathome19 committed:
Merge branch 'trunk' into trunk

danielathome19 committed:
Merge branch 'trunk' into trunk

zygoloid commented on explorer/ast/static_scope.cpp:
```diff
@@ -35,20 +35,21 @@ auto StaticScope::Add(std::string_view name, ValueNodeView entity,
 }
 
-void StaticScope::Print(llvm::raw_ostream& out) const {
+template<typename Action>
+void StaticScope::PrintCommon(Action action) const {
```
I don't think this will compile -- there's no corresponding member declared in class `StaticScope`. Is it possible that you forgot to commit the changes to `static_scope.h`?

danielathome19 replied:
That's right -- I've gone back and updated the header file in the PR.

danielathome19 committed:
Update static_scope.h -- added method header for `PrintCommon` function

danielathome19 committed:
Merge branch 'trunk' into trunk

danielathome19 committed:
Merge branch 'trunk' into trunk

zygoloid commented:
Thank you!

zygoloid commented on explorer/ast/static_scope.cpp:
```diff
@@ -35,20 +35,21 @@ auto StaticScope::Add(std::string_view name, ValueNodeView entity,
 }
 
-void StaticScope::Print(llvm::raw_ostream& out) const {
+template<typename Action>
```
```suggestion
template <typename Action>
```

zygoloid commented on explorer/ast/static_scope.cpp:
```diff
@@ -136,20 +137,14 @@ auto StaticScope::TryResolveHere(std::string_view name,
+         << "'" << name
+         << (it->second.status == NameStatus::KnownButNotDeclared
+                 ? "' has not been declared yet"
+                 : "' is not usable until after it has been completely declared");
```
```suggestion
                 : "' is not usable until after it has been completely "
                   "declared");
```

zygoloid commented on explorer/ast/static_scope.h:
```diff
@@ -55,6 +55,9 @@ class StaticScope {
   auto Add(std::string_view name, ValueNodeView entity,
            NameStatus status = NameStatus::Usable) -> ErrorOr<Success>;
 
+  template<typename Action>
```
```suggestion
  template <typename Action>
```

zygoloid committed:
Apply formatting changes from clang-format.
